src/dask_awkward/lib/io/io.py

```diff
@@ -620,16 +625,34 @@ def from_map(
             packed=packed,
         )
 
-    # Special `io_func` implementations can implement mocking and optionally
-    # support buffer projection.
-    if io_func_implements_mocking(func):
+    kw = {}
+    if io_func_implements_projection(func):
+        # Special `io_func` implementations can do buffer projection - choosing columns
+        # so here we start with a blank report
         io_func = func
-        array_meta = cast(ImplementsMocking, func).mock()
-    # If we know the meta, we can spoof mocking
-    elif meta is not None:
-        io_func = IOFunctionWithMocking(meta, func)
-        array_meta = meta
+        array_meta, report = typetracer_with_report(
+            form_with_unique_keys(io_func.form, "@"),
+            highlevel=True,
+            behavior=io_func.behavior,
+            buffer_key=render_buffer_key,
+        )
```

Comment:
This is the big change:

Control over buffer names now solely rests with dask-awkward, rather than the provider of the IO function e.g. uproot. 

This means that dask-awkward generates buffer names and gives them to the IO-function, meaning that the IO-functions now need to understand how to parse these keys instead of dask-awkward asking the IO-functions to parse them. This is all semantics; this is a trivial operation and we can define public functions in dask-awkward to do this.

Defining our column-optimisation interface at the Form level still ensures unique names, but it now requires that the IO-function understands how to map from the form to the buffers it works.[^vm] A nice perk of this change is that because dask-awkward now controls buffer names, we can ensure that they are globally unique (i.e. between different input layers) e.g. by embedding the layer ID. This might prove useful.

[^vm]: To understand the consequence of this, I need to more intimately remind myself how coffea currently remaps keys (where keys are VM instructions) to that which it presents to uproot.